### ISSUE TITLE
Module: content_extractor

### ISSUE THREAD

**Issue by oliversauter** [WorldBrain/crawler#4]
## Goal:

Extracting the important content bits from the HTML code we saved for each article URL

## Input

Gets the HTML file for each URL in table `articles` where column `processed_content_extractor`=`false` from django app

## Output

Gives back a JSON file containing
 -  `domain_name`, 
 - `url`,
 -  `text`,
 -  `keywords`(list),
 -  `authors`(list),
 -  `publish_date`, 
 - `tags`(list), 
 - `summary`,
 -  `links`(list)
 - `parse_time``

 to the django app, which saves it to the respective columns in table `articles`.

## Technical Implementation

I think you guys already have built the extractor we can use to parse the elements out of the HTML file, right? @lwm @arsenal9971

**Comment by oliversauter** [comment]
Here we could use different parsers, important is that we can change the input source.
[Newspaper3k](https://github.com/codelucas/newspaper) has a pretty well build dom extractor, but it might be tricky to use it without a http source.

Other options are beautifulsoup or boilerpipe

**Comment by arpitgogia** [comment]
A quick Google Search on **article content extraction python** reveals two popular libraries [Goose](https://github.com/grangier/python-goose) and [Newspaper](https://github.com/codelucas/newspaper). Judging only by the number of stars, we can go for Newspaper. Just a thought, I can perform a dry run on a number of articles and see how each library performs. 
Another question I have is the same as the last issue, will this be a sub-module of the Django app or an independent module.

**Comment by oliversauter** [comment]
here you also have a comparison of boilerpipe and goose: https://gist.github.com/eldilibra/5637215

**Comment by arpitgogia** [comment]
Its a win for Boilerpipe as compared to Goose. However, the Python Library for Boilerpipe is a wrapper over the original Java Library, it's out of development and of course requires Java as a dependency. I think Newspaper is the way to go.

**Comment by oliversauter** [comment]
yes, it is called newspaper3k. We already played with it a little bit. it normally takes a URL as input, which it parses and then extracts. 

The question is, if we can just give it a HTML file or somehow fake the URL to be from our DB?

**Comment by arpitgogia** [comment]
I browsed through the code for Newspaper3k and it appears that the Objects that we need to use to parse a URL have constructors with a URL field only and no support for raw HTML. 
One way we can do this is to create a separate endpoint for each article HTML stored in the DB. We can supply that URL to Newspaper3k and it should function the same way as providing an absolute link to an article. The content extractor will thus be independently deployed from the Django app. The Django app will have to have another API that can take meta data extracted by the content-parser and save it in the DB (as already mentioned above).

**Comment by oliversauter** [comment]
ok sounds like a plan. Let's do it like that. 

In case we want to extract other elements than provided by newspaper, is that possible to modify NP or are we limited to what newspaper3k makes possible?

**Comment by arpitgogia** [comment]
Their main extraction logic is available [here](https://github.com/codelucas/newspaper/blob/master/newspaper/extractors.py). It looks like a lot of Regular Expression matching as well as HTML tag based data extraction is used. We can definitely modify it but it would take time to identify more patterns among articles and then test the new build. 
Have you tried the [demo](http://newspaper-demo.herokuapp.com/)?

**Comment by oliversauter** [comment]
@hoshomoh @tobisanya @arpitgogia 

I would say we discuss the in or outside of django issue here, since it belongs to this issue:

**Question of Arpit:**
> Another question I have is the same as the last issue, will this be a sub-module of the Django app or an independent module?

**Comment by hoshomoh** [comment]
@arpitgogia can you go over how you would do this if it were to be a sub-module or an independent module

**Comment by arpitgogia** [comment]
Independent Module
-

1. Deployed separately from the Django app.
2. Periodically calls an API at Django app that returns a dummy URL pointing to the article HTML from the table.
3. Retrieves and parses the HTML. Sends it back to the Django app using another API which then updates the database.

Sub-Module
-
1. Use Celery to periodically execute this as a task.
2. Retrieve the article URL directly from the database, parse the article and update the table accordingly.

**Comment by hoshomoh** [comment]
I think it's best we go with sub-module because we can pretty much configure when and how the celery task are executed and the there would be 1 to none API call and the output can be directly written to the ```articles``` table. And in terms of scalability, if for any reason this affects the performance of other API'S we can move this to a seperate node (DevOps guys would understand this better). @arpitgogia @tobisanya  What do you guys think?

**Comment by oliversauter** [comment]
@arpitgogia when you say:
>Retrieve the article URL directly from the database, parse the article and update the table accordingly.

Do you mean reparsing it with newspaper3k? Then we would not need to save the HTML file with the Scrapy crawler, right?

If that is the case, I am not sure, if it is the right path to go. I find a solution where we have to visit a website twice in order to have it in our database rather suboptimal, since spidering a website already takes quite long with scrapy. Visiting all the articles again and parsing it would unnecessarily extend the time to have a site truly indexed + then we would have to develop the `article_detector` afterall, because we would for sure have to parse 30-50% unnecessary pages (like overview pages etc.)

**Comment by arpitgogia** [comment]
I did mean that when I mentioned it. And actually I realised that it wouldn't be a desirable way to accomplish this. But if it has to be made a sub-module, that is the only way I could think of.

**Comment by oliversauter** [comment]
How much more difficult would the API approach be? What are the downsides of it?

**Comment by arpitgogia** [comment]
The downsides as far as I can think are from the separate deployment. Extra management and any other related costs etc. In the long run, this may not be that tedious.

**Comment by oliversauter** [comment]
Ok, i assume in the long run, recrawling could be the more cost intensive part.

Do you estimate the extra mgmt to be intensive? What tasks would that extra mgmt include?

**Comment by hoshomoh** [comment]
I don't think we need to reparse the HTML with **newspaper3k** . If we already have the HTML stored from scrapy as agreed. [Goose here](https://github.com/grangier/python-goose) Supports extracting the metadata form HTML'S. So why not use that instead?

**Comment by oliversauter** [comment]
You @arpitgogia looked into it before, right?

From what i see it is not as detailed as newspaper3k when it comes to the extractable information.

We definitely need in the mid term stuff like author, keywords etc.

Goose only describes "meta tags" as abailable information besides title and text.

**Comment by arpitgogia** [comment]
Newspaper3k had better meta extraction. As mentioned in the above comments, Goose lost to another Library called Boilerpipe in benchmarks. I wasn't aware that Goose supports raw HTML based extraction as well. I can't seem to find any documentation for Python-Goose.

**Comment by oliversauter** [comment]
And @hoshomoh, yes we need to, since that part is baked into newspapwr3k

We could modify NP3K probably to just have the extractor part. But that would make it hard to benefit from their updates

This circumstance is the reason for the API idea.

**Comment by hoshomoh** [comment]
I just dug into the code now and I noticed that NP3K might actually accept HTML. [Check Here](https://github.com/codelucas/newspaper/blob/master/newspaper/article.py) Line 144. If you pass html to ```article.download()``` NP3K set that as the HTML, but if you don't then it would download it. I might be wrong, but then @arpitgogia @tobisanya check.

**Comment by arpitgogia** [comment]
I think @hoshomoh is right. I tried the following code:
```
from newspaper import Article
article = Article('')
file = open('test.html', 'r')
article.download(html=file.read())
```
And it worked! ```Test.html``` contained the HTML of this [article](http://www.collective-evolution.com/2016/09/07/the-impossible-quantum-space-engine-that-breaks-laws-of-physics-is-about-to-be-tested-in-space/). Then I tried to parse it directly through the URL and the results were same.
**In summary I think, this little hack will work**.

**Comment by hoshomoh** [comment]
I tested too and it worked. It might not be in their documentation. But, I am sure they had that in mind. So, @oliversauter we can actually send the HTML to NP3K.

**Comment by oliversauter** [comment]
Woooohooo great!

**Comment by arpitgogia** [comment]
So then we can conveniently make this as a sub module of the Cortex and use article HTML directly from the database. In that case, will this issue be moved to that repo?

**Comment by oliversauter** [comment]
Yes, I am on the road. Can you open it? Also adding some more details about the execution, which can be used for the readme later.
